app/models/surveys/survey.rb

```diff
@@ -86,13 +93,54 @@ def response(user)
     end
   end
 
-  def course_for(user)
-    @survey_assignment_ids ||= SurveyAssignment.where(survey_id: id).pluck(:id)
-    notification = user.survey_notifications
-                       .where(survey_assignment_id: @survey_assignment_ids)
-                       .first
-    # If there's no course from a notification, fall back to the user's latest course
-    notification&.course_id ? Course.find(notification.course_id).slug : user.courses.last&.slug
+  def course_for
+    load_survey_assignment_ids
+    notifications_by_user = fetch_notifications_by_user
+    course_ids = notifications_by_user.values.map(&:course_id).uniq
+    courses_by_id = load_courses_by_id(course_ids)
+    assign_survey_courses_to_users(notifications_by_user, courses_by_id)
+  end
+
+  def load_survey_assignment_ids
+    @survey_assignment_ids = SurveyAssignment.where(survey_id: id).pluck(:id)
+  end
+
+  def fetch_notifications_by_user
+    SurveyNotification
+      .joins(:courses_user)
+      .where(
+        courses_users: { user_id: @users_by_id.keys },
+        survey_assignment_id: @survey_assignment_ids
+      )
+      .select(
+        'survey_notifications.id',
+        'survey_notifications.course_id',
+        'courses_users.user_id AS user_id'
+      )
+      .order('survey_notifications.id')
+      .group_by(&:user_id)
+      .transform_values(&:first)
+  end
+
+  def load_courses_by_id(course_ids)
+    Course.where(id: course_ids).select(:id, :slug).index_by(&:id)
+  end
+
+  def assign_survey_courses_to_users(notifications_by_user, courses_by_id)
+    notifications_by_user.each do |user_id, notification|
+      user = @users_by_id[user_id]
+      course = courses_by_id[notification.course_id] || fallback_course_for_user(user_id)
+      user.define_singleton_method(:survey_course) { course }
```

Comment:
I think I understand what is going on here with `define_singleton_method`, but there must be a more elegant way to insert the course into the `notifications_by_user` data structure.

Some comments explaining the structure of this object (probably above `fetch_notifications_by_user`, since that's where the structure is created) would be helpful.